File: python/src/multiagent_boilerplate/artifacts/store.py
"""The seam for artifact persistence. LocalArtifactStore is the shipped default; implement the
same two-method ArtifactStore protocol against S3/GCS/a database and pass it to the orchestrator
to swap backends (see docs/extending.md). Callers only ever see an artifact reference dict,
never a storage-specific type (notes section 4-5)."""

from __future__ import annotations

import json
import uuid
from datetime import UTC, datetime
from pathlib import Path
from typing import Any, Protocol, runtime_checkable

from ..harness.validate import validate_artifact_ref


@runtime_checkable
class ArtifactStore(Protocol):
    async def write(self, kind: str, summary: str, content: Any, created_by: str) -> dict: ...
    async def read(self, artifact_id: str) -> Any: ...


class LocalArtifactStore:
    def __init__(self, root_dir: str | Path) -> None:
        self._root_dir = Path(root_dir)

    async def write(self, kind: str, summary: str, content: Any, created_by: str) -> dict:
        self._root_dir.mkdir(parents=True, exist_ok=True)
        artifact_id = str(uuid.uuid4())
        path = self._root_dir / f"{artifact_id}.json"
        serialized = json.dumps(content, indent=2)
        path.write_text(serialized, encoding="utf-8")

        ref = {
            "artifactId": artifact_id,
            "kind": kind,
            "sizeBytes": len(serialized.encode("utf-8")),
            "createdBy": created_by,
            "createdAt": datetime.now(UTC).isoformat(),
            "summary": summary,
            "uri": f"file://{path}",
        }
        validate_artifact_ref(ref)
        return ref

    async def read(self, artifact_id: str) -> Any:
        path = self._root_dir / f"{artifact_id}.json"
        return json.loads(path.read_text(encoding="utf-8"))

File: typescript/src/artifacts/store.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { randomUUID } from "node:crypto";
import { resolve } from "node:path";
import { validateArtifactRef } from "../harness/validate.js";

export interface ArtifactRef {
  artifactId: string;
  kind: string;
  sizeBytes: number;
  createdBy: string;
  createdAt: string;
  summary: string;
  uri: string;
}

export interface WriteArtifactInput {
  kind: string;
  summary: string;
  content: unknown;
  createdBy: string;
}

/**
 * The seam for artifact persistence. `LocalArtifactStore` is the shipped default; implement
 * this same two-method interface against S3/GCS/a database and pass it to the orchestrator to
 * swap backends (see docs/extending.md). Callers only ever see an ArtifactRef, never a
 * storage-specific type (notes section 4-5), so nothing above this line changes.
 */
export interface ArtifactStore {
  write(input: WriteArtifactInput): Promise<ArtifactRef>;
  read(artifactId: string): Promise<unknown>;
}

/** Local filesystem artifact store (default). */
export class LocalArtifactStore implements ArtifactStore {
  constructor(private readonly rootDir: string) {}

  async write(input: WriteArtifactInput): Promise<ArtifactRef> {
    await mkdir(this.rootDir, { recursive: true });
    const artifactId = randomUUID();
    const path = resolve(this.rootDir, `${artifactId}.json`);
    const serialized = JSON.stringify(input.content, null, 2);
    await writeFile(path, serialized, "utf-8");

    const ref: ArtifactRef = {
      artifactId,
      kind: input.kind,
      sizeBytes: Buffer.byteLength(serialized, "utf-8"),
      createdBy: input.createdBy,
      createdAt: new Date().toISOString(),
      summary: input.summary,
      uri: `file://${path}`,
    };
    validateArtifactRef(ref);
    return ref;
  }

  async read(artifactId: string): Promise<unknown> {
    const path = resolve(this.rootDir, `${artifactId}.json`);
    const raw = await readFile(path, "utf-8");
    return JSON.parse(raw) as unknown;
  }
}
